util/src/main/java/org/datacommons/util/CoordinatesResolver.java

```diff
@@ -0,0 +1,112 @@
+import org.datacommons.proto.Recon.ResolveCoordinatesResponse;
+
+/** Resolves nodes with lat-lngs by calling the DC coordinates resolution API. */
+public class CoordinatesResolver {
+  private final AtomicBoolean resolved = new AtomicBoolean(false);
+
+  private final Set<Coordinate> resolveCoordinates = ConcurrentHashMap.newKeySet();
+
+  private final ConcurrentHashMap<Coordinate, Set<String>> resolvedCoordinates =
+      new ConcurrentHashMap<>();
+
+  private final ReconClient client;
+
+  public CoordinatesResolver(ReconClient client) {
+    this.client = client;
+  }
+
+  public void submitNode(PropertyValues node) {
+    if (resolved.get()) {
+      throw new IllegalStateException("submitNode called after remote resolution.");
+    }
+    getCoordinate(node).ifPresent(resolveCoordinates::add);
+  }
+
+  // TODO: Pick the ID based on a preferred list.
+  public String resolveNode(PropertyValues node) {
+    return getCoordinate(node)
+        .filter(resolvedCoordinates::containsKey)
+        .flatMap(coordinate -> resolvedCoordinates.get(coordinate).stream().findFirst())
+        .orElse("");
+  }
+
+  // TODO: Support chunking in batches of max size 500.
+  public void remoteResolve() throws IOException, InterruptedException {
+    if (resolved.get()) {
+      throw new IllegalStateException("remoteResolve called after remote resolution.");
+    }
+    resolved.set(true);
+
+    if (resolveCoordinates.isEmpty()) {
+      return;
+    }
+
+    ResolveCoordinatesRequest request =
+        ResolveCoordinatesRequest.newBuilder().addAllCoordinates(resolveCoordinates).build();
+    ResolveCoordinatesResponse response = client.resolveCoordinates(request);
+    populateResolvedCandidates(response);
+  }
+
+  boolean isResolved() {
+    return resolved.get();
+  }
+
+  private void populateResolvedCandidates(ResolveCoordinatesResponse response) {
+    response
+        .getPlaceCoordinatesList()
+        .forEach(
+            placeCoordinate -> {
+              if (placeCoordinate.getPlaceDcidsCount() > 0) {
+                resolvedCoordinates.put(
+                    Coordinate.newBuilder()
+                        .setLatitude(placeCoordinate.getLatitude())
+                        .setLongitude(placeCoordinate.getLongitude())
+                        .build(),
+                    new LinkedHashSet<>(placeCoordinate.getPlaceDcidsList()));
+              }
+            });
+  }
+
+  private static Optional<Coordinate> getCoordinate(PropertyValues node) {
+    if (node.containsPvs(Vocabulary.LATITUDE) && node.containsPvs(Vocabulary.LONGITUDE)) {
+
+      Optional<Double> optLat = getDoubleValue(node.getPvsMap().get(Vocabulary.LATITUDE));
+      Optional<Double> optLng = getDoubleValue(node.getPvsMap().get(Vocabulary.LONGITUDE));
+
+      if (optLat.isPresent() && optLng.isPresent()) {
+        double lat = optLat.get();
+        double lng = optLng.get();
+
+        if (!Double.isNaN(lat) && !Double.isNaN(lng)) {
```

Comment:
We should do some sanity checks on the lat/lng?  Add TODO?

Reply:
There's a TODO on the method below (`getDoubleValue`) to support other formats.